### PR TITLE
Improve .dyno library files

### PR DESCRIPTION
**Pull request by mppf**
This PR improves the prototype `.dyno` file format. It is a continuation of PRs #23696 #23697 #23708 #23711.

This PR makes the following adjustments:
 * Introduces a `libraries` namespace and subdirectory to contain most of the library support. Uses a separate subdirectory since it's a cross-cutting concern and the existing categories don't really seem appropriate.
 * Migrates library reading and writing code to supporting types `LibraryFile` and `LibraryFileWriter`. The `LibraryFile` type uses `mmap` to make it easy to access additional parts of a `LibraryFile` without requiring the whole file be read. This is particularly useful for the locations section, which is not yet implemented, but which I expect to be large but rarely used.
 * Switches serialized `UniqueString` handling directly store strings shorter than `LONG_STRING_SIZE` (currently `20`) and to store long string indices for any strings longer than that. These long string indices point into a long strings section in the file format.
 * Changes the `Deserializer` to deserialize from a memory region now that we are using `mmap`. This is still OK for testing because they were using `stringstream` anyway.
 * Describes the library file format in a new document `frontend/docs/file-format.rst` and also in some `struct` definitions in `LibraryFileFormat.h`
 * Adds new C++ testing for the library files in `testLibraryFile.cpp`
 * To enable `--dyno-verify-serialization` to continue to work while `.dyno` files do not store serialized comments, added `Context::Configuration::includeComments` to configure whether or not some comments should be included in the uAST.

Future Work:
 * implement the described Locations section. Currently, the `.dyno` files don't include locations at all.

Reviewed by @benharsh - thanks!

- [x] full comm=none testing

### CODE REVIEW

File: frontend/lib/framework/UniqueString.cpp [hunk 1]
```diff
@@ -200,24 +201,49 @@ std::ostream& operator<<(std::ostream& os, const chpl::UniqueString& uStr) {
...
+  } else {
+    // store the string through the long strings table
+    uint32_t uid = ser.cacheString(c_str(), len);
+    // won't work if the high bit is set
+    CHPL_ASSERT((uid >> 31) == 0);
+    // compute the uid in big-endian order with high bit set
+    uint32_t num = htonl(uid);
+    uint8_t bytes[4];
+    memcpy(&bytes[0], &num, sizeof(num));
+    bytes[0] |= 0x80; // set the high bit
```
Comment (benharsh) [inline frontend/lib/framework/UniqueString.cpp:220]: Would these kinds of operations and checks read more easily as a macro?

File: frontend/lib/libraries/LibraryFile.cpp [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+      uAstOffset+sizeof(AstSectionHeader) > len) {
+    context->error(Location(), "Invalid module section header in %s",
+                   libPath.c_str());
+    return false;
+  }
+
+  const AstSectionHeader* astHdr =
+    (const AstSectionHeader*) (data + uAstOffset);
+
+  if (astHdr->magic != 0x0003bb1e5ec110e0 ||
```
Comment (benharsh) [inline frontend/lib/libraries/LibraryFile.cpp:208]: I think this should use the declared constant

File: frontend/doc/file-format.rst [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+The uAST section consists of serialized uAST entries in the order they
+appear within the file. The top-level entry will be the uAST for the
+corresponding module. Note that the symbol table can point to an
+individual entry within the serialized uAST.
+
+IDs are not stored here. They are recomputed when the uAST is read.
+
+The uAST section consists of:
+
+ * 8 bytes of magic number 0x0003bb1e5ec110e0
```
Comment (benharsh) [inline frontend/doc/file-format.rst:178]: This appears to be the same magic number as the previous section, and would need to be fixed in the code as well.

File: frontend/lib/framework/UniqueString.cpp (hunk 1 above)
Comment (mppf) [reply frontend/lib/framework/UniqueString.cpp:220]: What did you have in mind? IMO this is small enough and is contained to `UniqueString::serialize` and `UniqueString::deserialize`, so I'm not sure a macro would make it easier to read or maintain.

File: frontend/include/chpl/libraries/LibraryFile.h [hunk 4]
```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
```
[the remark this answers is not in the document]
Comment (mppf) [reply frontend/include/chpl/libraries/LibraryFile.h]: Good point. I've created PR #23734 to fix the script checking these.

File: frontend/lib/framework/UniqueString.cpp (hunk 1 above)
Comment (benharsh) [reply frontend/lib/framework/UniqueString.cpp:220]: I was just thinking something like ``SET_HIGH(bytes[0])``, to eliminate the need for the comment. Anyways, if we don't expect or intend to need this much elsewhere I think it's fine to leave as-is.

**Review by benharsh — APPROVED** [review]
Other than the magic number literal, this looks good to go!

**Comment by mppf** [comment]
I did a quick experiment with compiling `examples/hello.chpl` with 1.32 vs after this PR with and without a pre-parsed standard library saved in a `.dyno` file:

1.32
 * with regular (whole program) compilation: `parseAndConvertUast :   0.236 seconds`
 * with `chpl_standard.dyno`: `parseAndConvertUast :   0.241 seconds`
   * `chpl_stardard.dyno` size: 21M

after this PR
 * with regular (whole program) compilation: `parseAndConvertUast :   0.251 seconds`
 * with chpl_standard.dyno: `parseAndConvertUast :   0.162 seconds`
   * `chpl_stardard.dyno` size: 2.2M

So there is some performance improvement on the order of 0.1s. We still need to add location information to the `.dyno` file but I am not expecting that to change this performance because we are `mmap`ing the file and will just not read that part unless emitting an error message.

**Pull request merged**